File: app/src/main/java/io/github/lazymockbean/parser/LazyMockDefinitionParser.java
package io.github.lazymockbean.parser;

import io.github.lazymockbean.annotation.LazyInjectMockBeans;
import io.github.lazymockbean.data.BeanDependencyFinder;
import io.github.lazymockbean.data.LazyMockDefinition;
import io.github.lazymockbean.data.LazyMockTarget;
import io.github.lazymockbean.utils.Annotations;
import io.github.lazymockbean.utils.MockGenerator;
import io.github.lazymockbean.utils.SpringBeans;
import lombok.experimental.UtilityClass;
import org.apache.commons.lang3.ArrayUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.TestContext;
import org.springframework.util.ClassUtils;
import org.springframework.util.ReflectionUtils;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@UtilityClass
public class LazyMockDefinitionParser {
    private static final List<String> EXCLUDE_BEAN_PACKAGE = Arrays.asList("java.util", "java.lang");

    public Set<LazyMockDefinition> parse(TestContext testContext) {
        return Arrays.stream(testContext.getTestClass().getDeclaredFields())
                .filter(Annotations::hasLazyMockBeanAnotations)
                .flatMap(mockingField -> {
                    Class<?> targetFieldType = mockingField.getType();
                    Object mock = MockGenerator.generate(mockingField, testContext, targetFieldType);
                    Class<?>[] injectTargets = Annotations.getInjectTargets(mockingField);
                    if (!ArrayUtils.isEmpty(injectTargets)) {
                        return parseSpecificTargets(injectTargets, testContext, mockingField, mock).stream();
                    } else {
                        return parseAutowiredDependencyBeans(testContext, mockingField, mock).stream();
                    }
                }).collect(Collectors.toSet());
    }

    private List<LazyMockDefinition> parseSpecificTargets(
            Class<?>[] injectTargets, TestContext testContext, Field mockingField, Object mockObject
    ) {
        return Arrays.stream(injectTargets)
                .flatMap(it -> generateDefinition(testContext, it, mockingField, mockObject).stream())
                .collect(Collectors.toList());
    }

    private List<LazyMockDefinition> generateDefinition(
            TestContext testContext, Class<?> target, Field mockingField, Object mockObject
    ) {
        return findMockingTarget(testContext, target, mockingField.getType()).stream()
                .map(it -> {
                    Object targetBean = it.getTargetBean();
                    Field targetField = it.getTargetField();
                    Object originValue = getFieldValue(targetBean, targetField);
                    return new LazyMockDefinition(mockingField, targetField, targetBean, originValue, mockObject);
                }).collect(Collectors.toList());
    }

    private Object getFieldValue(Object targetBean, Field targetField) {
        try {
            targetField.trySetAccessible();
            return targetField.get(targetBean);
        } catch (IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    private List<LazyMockTarget> findMockingTarget(TestContext testContext, Class<?> targetObject, Class<?> targetFieldType) {
        List<Object> beans = SpringBeans.findAllBeansWithoutProxy(testContext, targetObject);
        return beans.stream()
                .map(bean -> {
                    Field targetFieldInBean = ReflectionUtils.findField(bean.getClass(), null, targetFieldType);
                    if (targetFieldInBean == null) {
                        String message = String.format("not exist field [%s] in bean [%s]", targetFieldType.getName(), targetObject.getSimpleName());
                        throw new IllegalStateException(message);
                    }
                    return new LazyMockTarget(bean, targetFieldInBean);
                }).collect(Collectors.toList());
    }

    private List<LazyMockDefinition> parseAutowiredDependencyBeans(TestContext testContext, Field mockingField, Object mockObject) {
        Class<?> targetFieldType = mockingField.getType();
        BeanDependencyFinder beanDependencyFinder = new BeanDependencyFinder();

        Arrays.stream(testContext.getTestClass().getDeclaredFields())
                .filter(it -> it.isAnnotationPresent(LazyInjectMockBeans.class))
                .forEach(it -> beanDependencyFinder.push(it.getType()));

        if (beanDependencyFinder.isEmpty()) {
            Arrays.stream(testContext.getTestClass().getDeclaredFields())
                    .filter(it -> it.isAnnotationPresent(Autowired.class))
                    .forEach(it -> beanDependencyFinder.push(it.getType()));
        }

        List<LazyMockDefinition> mockDefinitions = new ArrayList<>();

        while (beanDependencyFinder.hasMore()) {
            Class<?> dependentClazz = beanDependencyFinder.pop();
            List<Object> beans = SpringBeans.findAllBeansWithoutProxy(testContext, dependentClazz);
            for (Object bean : beans) {
                List<LazyMockDefinition> definitions = Arrays.stream(bean.getClass().getDeclaredFields())
                        .filter(it -> it.getType() == targetFieldType)
                        .flatMap(it -> generateDefinition(testContext, dependentClazz, mockingField, mockObject).stream())
                        .collect(Collectors.toList());
                mockDefinitions.addAll(definitions);

                Arrays.stream(bean.getClass().getDeclaredFields())
                        .filter(it -> isValidBeanType(it.getType()))
                        .filter(it -> SpringBeans.hasAnyBean(testContext, it.getType()))
                        .forEach(it -> beanDependencyFinder.push(it.getType()));
            }
        }
        return mockDefinitions;
    }

    private Boolean isValidBeanType(Class<?> type) {
        if (type == null || type.getPackage() == null) {
            return false;
        }

        return !ClassUtils.isPrimitiveOrWrapper(type) &&
                !ClassUtils.isPrimitiveArray(type) &&
                !ClassUtils.isPrimitiveWrapperArray(type) &&
                !EXCLUDE_BEAN_PACKAGE.contains(type.getPackage().toString());
    }

}

File: app/src/main/kotlin/io/github/taesk/lazymockbean/parser/LazyMockDefinitionParser.kt
package io.github.taesk.lazymockbean.parser

import io.github.taesk.lazymockbean.annotation.LazyInjectMockBeans
import io.github.taesk.lazymockbean.data.DependencyFinder
import io.github.taesk.lazymockbean.data.LazyMockDefinition
import io.github.taesk.lazymockbean.data.LazyMockTarget
import io.github.taesk.lazymockbean.utils.Annotations
import io.github.taesk.lazymockbean.utils.MockGenerator
import io.github.taesk.lazymockbean.utils.SpringBeans
import org.springframework.test.context.TestContext
import org.springframework.util.ClassUtils
import org.springframework.util.ReflectionUtils
import java.lang.reflect.Field
import kotlin.reflect.KClass

object LazyMockDefinitionParser {
    private val EXCLUDE_BEAN_PACKAGE: List<String> = listOf("java.util", "java.lang")

    fun parse(testContext: TestContext): Set<LazyMockDefinition> {
        return testContext.testClass.declaredFields
            .filter { Annotations.hasLazyMockBeanAnotations(it) }
            .flatMap { mockingField ->
                val targetFieldType = mockingField.type
                val mock = MockGenerator.generate(mockingField, testContext, targetFieldType)
                val injectTargets = Annotations.getInjectTargets(mockingField)
                if (injectTargets.isNotEmpty()) {
                    parseSpecificTargets(injectTargets, testContext, mockingField, mock)
                } else {
                    parseAutowiredDependencyBeans(testContext, mockingField, mock)
                }
            }.toSet()
    }

    private fun parseSpecificTargets(
        injectTargets: Array<KClass<*>>,
        testContext: TestContext,
        mockingField: Field,
        mockObject: Any
    ): List<LazyMockDefinition> {
        return injectTargets.flatMap { generateDefinition(testContext, it.java, mockingField, mockObject) }
    }

    private fun generateDefinition(
        testContext: TestContext,
        target: Class<*>,
        mockingField: Field,
        mockObject: Any
    ): List<LazyMockDefinition> {
        return findMockingTarget(
            testContext = testContext,
            targetObject = target,
            targetFieldType = mockingField.type
        ).map { (targetBean, targetField) ->
            val originValue = targetField.getForce(targetBean)
            LazyMockDefinition(
                mockingField = mockingField,
                targetField = targetField,
                targetBean = targetBean,
                origin = originValue,
                mock = mockObject,
            )
        }
    }

    private fun findMockingTarget(
        testContext: TestContext,
        targetObject: Class<*>,
        targetFieldType: Class<*>
    ): List<LazyMockTarget> {
        val beans = SpringBeans.findAllBeansWithoutProxy(testContext, targetObject)
        return beans.map { bean ->
            val targetFieldInBean = ReflectionUtils.findField(bean::class.java, null, targetFieldType)
            requireNotNull(targetFieldInBean) { "not exist field [${targetFieldType.name}] in bean [${targetObject.simpleName}]" }
            LazyMockTarget(
                targetBean = bean,
                targetField = targetFieldInBean
            )
        }
    }

    private fun Field.getForce(parent: Any): Any {
        this.trySetAccessible()
        return this.get(parent)
    }

    private fun parseAutowiredDependencyBeans(
        testContext: TestContext,
        mockingField: Field,
        mockObject: Any,
    ): List<LazyMockDefinition> {
        val targetFieldType = mockingField.type
        val dependencyFinder = DependencyFinder()
        testContext.testClass.declaredFields
            .filter { it.isAnnotationPresent(LazyInjectMockBeans::class.java) }
            .forEach { dependencyFinder.push(it.type) }
        val mockDefinitions = mutableListOf<LazyMockDefinition>()
        while (!dependencyFinder.isEmpty()) {
            val depedentClazz = dependencyFinder.pop()
            val beans = SpringBeans.findAllBeansWithoutProxy(testContext, depedentClazz)
            for (bean in beans) {
                bean.javaClass.declaredFields
                    .filter { it.type == targetFieldType }
                    .flatMap { generateDefinition(testContext, depedentClazz, mockingField, mockObject) }
                    .let { mockDefinitions.addAll(it) }
                bean.javaClass.declaredFields
                    .filter { isValidateBeanType(it.type) }
                    .filter { SpringBeans.hasAnyBean(testContext, it.type) }
                    .filter { !dependencyFinder.hasAlradyFound(it.type) }
                    .forEach { dependencyFinder.push(it.type) }
            }
        }
        return mockDefinitions
    }

    private fun isValidateBeanType(type: Class<*>): Boolean {
        return !ClassUtils.isPrimitiveOrWrapper(type) &&
            !ClassUtils.isPrimitiveArray(type) &&
            !ClassUtils.isPrimitiveWrapperArray(type) &&
            type.`package`.toString() !in EXCLUDE_BEAN_PACKAGE
    }
}
